[None] [ci] Reorganize CMake and Python integration test infrastructure for C++ tests

Funatiq opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Tests**
  - Reorganized tests into unit and end-to-end groups; moved and rebalanced multi‑GPU, kernel, layer, runtime, batch-manager, and e2e coverage. Added and removed multiple test targets, adjusted discovery/invocation paths, added new multi‑GPU and e2e tests (including userBuffer), shortened some L0 timeouts, and changed several tests from silent skips to assertion failures.

- **Chores**
  - Centralized build-type propagation and unified Google‑tests build step; improved test logging, artifact backup, and CI test-path triggers.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up
-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

Cpp tests that don’t seem to run in ci:
* allReduceKernelTest
* allReduceFusionTest
* moeAllReduceFusionTest

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
chore: Remove redundant check for "gpt_session" in generate_excluded_test_list function in cpp_common.py

This check was redundant because GptSession has been removed.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[test] Refactor test structure by moving tests to unit_tests directory

- Move several test cases files from the tests directory to the unit_tests directory.
- Update CMakeLists.txt files to reflect the new directory structure.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[test] Move end-to-end tests to e2e_tests directory

- Move end-to-end test files for batch manager and executor to the e2e_tests directory.
- Introduce new CMakeLists.txt files to organize end-to-end tests for the batch manager and executor components.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[test] Enhance build configuration for C++ tests

- Introduced a new fixture to specify build type for C++ tests, defaulting to "Release".
- Updated existing fixtures to utilize the new build type parameter.
- Modified build commands to accommodate the specified build type, improving flexibility in test configurations.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A30-CPP-1, A30-CPP-2, A30-CPP-3"

tensorrt-cicd commented:
[PR_Github #14779](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14779/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #11158 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11158/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --stage-list "A30-CPP-1, A30-CPP-2, A30-CPP-3"

tensorrt-cicd commented:
[PR_Github #14808](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14808/) [ run ] triggered by Bot

Funatiq committed:
[test] Improve C++ test fixtures and backup handling

- Added documentation for the build_type and build_dir fixtures to clarify their purpose.
- Enhanced the keep_log_files fixture to use pathlib for backup directory creation and improved XML file handling.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A30-CPP-1, A30-CPP-2, A30-CPP-3"

tensorrt-cicd commented:
[PR_Github #14830](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14830/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14973](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14973/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11305 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11305/) completed with status: 'FAILURE'

Funatiq committed:
[ci] Update L0_MergeRequest.groovy to reflect the new test directory structure

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A30-CPP-1, A30-CPP-2, A30-CPP-3"

tensorrt-cicd commented:
[PR_Github #14996](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14996/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14996](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14996/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11325 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11325/) completed with status: 'SUCCESS'

Funatiq committed:
[test] Update test durations and timeout settings

- Updated test duration entries for C++ tests in .test_durations.
- Adjusted timeout settings for specific tests in l0_a30.yml to improve test reliability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #15111](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15111/) [ run ] triggered by Bot

Funatiq committed:
[test] Move multi-GPU unit tests to separate directory

- Moved multi-GPU test files to unit_tests/multi_gpu directory, including cacheTransceiverTest, mpiUtilsTest, and various allReduce tests.
- Updated CMakeLists.txt files to include the new multi-GPU test directories and ensure proper test organization.
- Updated test_multi_gpu.py to reflect the new test directory structure.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented on tests/integration/test_lists/test-db/l0_a30.yml:
```diff
@@ -38,13 +38,13 @@ l0_a30:
-  - cpp/test_unit_tests.py::test_unit_tests[layers-80] TIMEOUT (90)
+  - cpp/test_unit_tests.py::test_unit_tests[layers-80] TIMEOUT (60)
   - cpp/test_unit_tests.py::test_unit_tests[runtime-80]
   - cpp/test_unit_tests.py::test_unit_tests[thop-80]
   - cpp/test_unit_tests.py::test_unit_tests[utils-80]
-  - cpp/test_e2e.py::test_model[-gpt-80] TIMEOUT (90)
-  - cpp/test_e2e.py::test_model[-gpt_executor-80] TIMEOUT (90)
+  - cpp/test_e2e.py::test_model[-gpt_executor-80] TIMEOUT (60)
   - cpp/test_e2e.py::test_model[-gpt_tests-80]
+  - cpp/test_e2e.py::test_model[-gpt-80] TIMEOUT (60)
```
The timeouts are in minutes while the recorded runtimes are in seconds, so it works out.

Funatiq committed:
[test] Add user buffer unit test to multi-GPU tests

- Moved userBufferTest to unit_tests/multi_gpu directory.
- Updated CMakeLists.txt and test_multi_gpu.py to reflect the new test structure.
- Added a new test function run_user_buffer_tests to test_multi_gpu.py.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A30-CPP-1, A30-CPP-2, A30-CPP-3, H100_PCIe-CPP-1, H100_PCIe-CPP-2,  DGX_H100-4_GPUs-CPP-1"

tensorrt-cicd commented:
[PR_Github #15132](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15132/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #15135](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15135/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11428 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11428/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #15167](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15167/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11455 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11455/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #15222](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15222/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11497](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11497/) completed with status: 'FAILURE'

DomBrown reviewed (commented):
Generally looks good. WIll happily approve if my one comment turns out not to be an issue.

Funatiq committed:
[test] Refactor C++ test fixtures and improve build directory handling

- Updated the `keep_log_files` fixture to dynamically create backup directories based on the build directory name.
- Enhanced the `find_build_dir` function with detailed docstring and improved return logic for resolving the C++ build directory.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[chore] Fix typo in disaggExecutorTest

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #15792](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15792/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15792](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15792/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11870](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11870/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #15816](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15816/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15816](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15816/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #11890](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11890/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15885](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15885/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15885](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15885/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11938](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11938/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #15941](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15941/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16172](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16172/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #16253](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16253/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #12218](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12218/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/12218/test-results/rerun_report.html) for details.

chzblych commented on tests/integration/test_lists/test-db/l0_a30.yml:
```diff
@@ -38,13 +38,13 @@ l0_a30:
   - cpp/test_unit_tests.py::test_unit_tests[executor-80]
   - cpp/test_unit_tests.py::test_unit_tests[kernels-80]
-  - cpp/test_unit_tests.py::test_unit_tests[layers-80] TIMEOUT (90)
+  - cpp/test_unit_tests.py::test_unit_tests[layers-80] TIMEOUT (60)
```
@Funatiq `60 mins` is the default timeout setting in the CI pipeline. Adding `TIMEOUT (60)` doesn't change anything.

Funatiq committed:
fixup! [test] Update test durations and timeout settings

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot reuse-pipeline
